File: internal/mechanics/resurrection.go
package mechanics

import (
	"math"

	"github.com/slycrel/slycrel/internal/model"
)

// ResurrectUser applies death penalties and brings the character back to life.
// Original: ResurrectUser in General_Unit.cp
// Penalties: -2% TotalExperience, -2% SpendingExperience, -1 Fame, +1 Faith, -20% CoinsHand
func ResurrectUser(char *model.Character) {
	char.TotalExperience -= int64(math.Round(float64(char.TotalExperience) * 0.02))
	char.SpendingExperience -= int64(math.Round(float64(char.SpendingExperience) * 0.02))
	char.Fame--
	char.Faith++
	char.CoinsHand -= int64(math.Round(float64(char.CoinsHand) * 0.20))

	if char.TotalExperience < 0 {
		char.TotalExperience = 0
	}
	if char.SpendingExperience < 0 {
		char.SpendingExperience = 0
	}
	if char.CoinsHand < 0 {
		char.CoinsHand = 0
	}

	char.Alive = true
	char.HitPoints = char.MaxHP
}

// NewDayForUser resets daily limits and restores the character for a new day.
// Original: NewDayForUser in Sly_MainUnit.cp
func NewDayForUser(char *model.Character) {
	char.Psyche = char.MaxPsyche
	char.TotalExperience += 15
	char.SpendingExperience += 15
	char.Movement = RandBetween(
		char.Speed-char.Speed/5,
		char.Speed+char.Speed/5,
	)
	char.Exploration = model.MaxExploration
	char.Spars = model.MaxSpars

	if !char.Alive {
		ResurrectUser(char)
	} else {
		char.HitPoints = char.MaxHP
	}
}

File: clients/web/mechanics/resurrection.js
import { MAX_EXPLORATION, MAX_SPARS } from '../model/character.js';
import { randBetween } from './rand.js';

// Mirrors internal/mechanics/resurrection.go ResurrectUser:
// -2% TotalExperience, -2% SpendingExperience, -1 Fame, +1 Faith,
// -20% CoinsHand, then restore HP and revive.
export function resurrectUser(c) {
  c.totalExperience -= Math.round(c.totalExperience * 0.02);
  c.spendingExperience -= Math.round(c.spendingExperience * 0.02);
  c.fame -= 1;
  c.faith += 1;
  c.coinsHand -= Math.round(c.coinsHand * 0.20);
  if (c.totalExperience < 0) c.totalExperience = 0;
  if (c.spendingExperience < 0) c.spendingExperience = 0;
  if (c.coinsHand < 0) c.coinsHand = 0;
  c.alive = true;
  c.hitPoints = c.maxHP;
}

// Mirrors NewDayForUser — resets daily limits + heals + revives if dead.
export function newDayForUser(c) {
  c.psyche = c.maxPsyche;
  c.totalExperience += 15;
  c.spendingExperience += 15;
  const spread = Math.floor(c.speed / 5);
  c.movement = randBetween(c.speed - spread, c.speed + spread);
  c.exploration = MAX_EXPLORATION;
  c.spars = MAX_SPARS;
  if (!c.alive) {
    resurrectUser(c);
  } else {
    c.hitPoints = c.maxHP;
  }
}
